### ISSUE TITLE
Commit 4e02f1a causes '!ps' output to differ from in vim.

### ISSUE THREAD

**Issue by blankname** [neovim/neovim#8217]
- Vim (version: 8, patches: 1-1542) behaves differently? yes
- Operating system/version: Linux

### Actual behaviour
ps shows lots of unrelated processes

### Expected behaviour
ps only shows processes with the same SID

I've verified that the behavior was introduced in `4e02f1a`.

**Comment by justinmk** [comment]
Good catch, thanks for bisecting. Can you compare the behavior of gVim, if you have it?

Vim does do setsid() in some cases of `mch_call_shell_fork()` but it's miserable to follow the logic. That's why I also did it for Nvim's `os_system()`.

**Comment by blankname** [comment]
:!ps for regular vim looks like this:
```
  PID TTY          TIME CMD
26391 pts/3    00:00:00 fish
26419 pts/3    00:00:00 vim
26447 pts/3    00:00:00 fish
26456 pts/3    00:00:00 ps
```

:!ps for gVim looks like this:
```
  PID TTY          TIME CMD
26645 pts/3    00:00:00 fish
26656 pts/3    00:00:00 ps
```

My current use case for the old behavior is a hack to close both nvim and it's parent terminal.
```
let g:patricide = 'false'
noremap <silent> <C-q> :let g:patricide = 'true' \| qa<cr>

aug VimLeaving
	au!
    au VimLeave * call VimLeaving()
aug END

function! VimLeaving()
  " kill session leader
  if g:patricide == 'true'
    let l:sid = system('string trim (ps -o sid= | tail -n1)')[:-2]
    exe ':!kill ' . l:sid
  endif
endfunction
```

**Comment by blankname** [comment]
The same commit also seems to have broken this command:
```
:w !sudo tee % > /dev/null                                                                 
sudo: no tty present and no askpass program specified                                      
                                                                                           
shell returned 1  
```

**Comment by justinmk** [comment]
"Fixing" this as attempted in https://github.com/neovim/neovim/pull/8389 will regress https://github.com/neovim/neovim/pull/8107 for `system()`, `systemlist()`, and `:!`. So CTRL-C on `:echo system('sleep 30|sleep 30|sleep 30')` will leave orphan processes.

Vim's approach is:

1. block signals
2. fork()
3. unblock signals in the child
4. on CTRL-C, send SIGINT to the process-group id (parent (`vim`) ignores the signal, child and its descendants do not)

To do this in Nvim AFAICT would require adding a case to `process_spawn()` to work without using libuv. Possibly this could be implemented like how `src/nvim/os/pty_process_unix.c` implements `ProcessType.kProcessTypePty`. We could add a `ProcessType.kProcessTypeUnix` and special-case it.

I'm pretty sure I don't want to spend my time doing this. But if anyone else wants to, it's welcome. 

> My current use case for the old behavior is a hack to close both nvim and it's parent terminal.

@blankname You can use `nvim_get_proc()` to get Nvim's parent PID:

    :echo nvim_get_proc(getpid())
    {'ppid': 2250, 'name': 'nvim', 'pid': 7909}

Does that address your use-case? If that's not good enough, we could add session-id to the return value of `nvim_get_proc()`. 

> :w !sudo tee % > /dev/null
> sudo: no tty present and no askpass program specified  

That's https://github.com/neovim/neovim/issues/1716 . We need to address that by leveraging `:terminal` for `:!`, not by continuing to special-case Nvim-in-a-TTY.

Treat Nvim as a server, not a terminal application. Or use Vim.

**Comment by blankname** [comment]
@justinmk Thanks, `nvim_get_proc()` should work for my use case.
I'm on linux and #1716 didn't affect me until the mentioned commit. I'll switch to using the plugin `lambdalisue/suda.vim` mentioned in that issue.
Feel free to close.

**Comment by justinmk** [comment]
Tagged as `wontfix` for the reasons mentioned above, but  https://github.com/neovim/neovim/issues/8678 outlines a potential path to restoring the Vim-compatible behavior.
